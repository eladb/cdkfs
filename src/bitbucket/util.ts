

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}