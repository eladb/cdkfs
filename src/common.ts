export const PROJEN_RC = '.projenrc.js';
export const PROJEN_DIR = '.projen';
export const PROJEN_MARKER = '~~ Generated by ' + 'projen'; // we split into two so /this/ file does not match the marker

// eslint-disable-next-line @typescript-eslint/no-require-imports
export const PROJEN_VERSION = require('../package.json').version;
