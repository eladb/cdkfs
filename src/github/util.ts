export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
