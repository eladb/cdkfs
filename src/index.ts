export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';