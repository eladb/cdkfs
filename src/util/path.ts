/**
 * Helper function to format a path as dot notation regardless of how it
 * was handed in.
 *
 * @param path - can be formatted as "path", "./path", or "/path"
 * @returns "./path"
 */
export function formatPathAsDotNotation(path: string): string {
  if (path.startsWith(".")) {
    return path;
  }

  if (path.startsWith("/")) {
    return `.${path}`;
  }

  return `./${path}`;
}
