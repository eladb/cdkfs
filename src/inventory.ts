import * as path from 'path';
import * as fs from 'fs-extra';

// eslint-disable-next-line @typescript-eslint/no-require-imports
const decamelize = require('decamelize');
const PROJEN_MODULE_ROOT = path.join(__dirname, '..');
const PROJECT_BASE_FQN = 'projen.Project';

type JsiiTypes = { [name: string]: JsiiType };

export interface ProjectOption {
  path: string[];
  name: string;
  switch: string;
  type: string;
  parent: string;
  docs?: string;
  default?: string;
  optional?: boolean;
  deprecated?: boolean;
}

export interface ProjectType {
  moduleName: string;
  pjid: string;
  fqn: string;
  typename: string;
  options: ProjectOption[];
  docs?: string;
  docsurl: string;
}

interface JsiiType {
  name: string;
  assembly: string;
  kind: string;
  abstract?: boolean;
  base?: string;
  fqn: string;
  interfaces?: string[];
  initializer?: {
    parameters?: Array<{
      name: string;
      type?: { fqn?: string };
    }>;
  };
  properties?: Array<{
    name: string;
    docs: {
      summary?: string;
      default?: string;
      deprecated?: string;
      stability?: string;
    };
    optional?: boolean;
    type?: {
      primitive?: string;
      fqn?: string;
    };
  }>;
  docs?: {
    summary?: string;
    deprecated?: string;
    custom?: {
      pjid?: string;
    };
  };
}

/**
 * Returns a list of project types exported the modules defined in `moduleDirs`.
 * This list will always also include the built-in projen project types.
 * Modules without a .jsii manifest are skipped.
 *
 * @param moduleDirs A list of npm module directories
 */
export function discover(...moduleDirs: string[]) {
  const jsii: JsiiTypes = {};

  const discoverJsii = (dir: string) => {
    const jsiiFile = path.join(dir, '.jsii');
    if (!fs.existsSync(jsiiFile)) { return; } // no jsii manifest

    const manifest = fs.readJsonSync(jsiiFile);
    for (const [fqn, type] of Object.entries(manifest.types as JsiiTypes)) {
      jsii[fqn] = type;
    }
  };

  // read all .jsii manifests from all modules (incl. projen itself) and merge
  // them all into a single map of fqn->type.
  for (const dir of [...moduleDirs, PROJEN_MODULE_ROOT]) {
    discoverJsii(dir);

    if (dir.includes('@') && fs.lstatSync(dir).isDirectory()) {
      const childDirs = fs.readdirSync(dir).map(file => path.join(dir, file));
      for (const child of childDirs) {
        discoverJsii(child);
      }
    }
  }

  const result = new Array<ProjectType>();

  for (const [fqn, typeinfo] of Object.entries(jsii)) {
    if (!isProjectType(jsii, fqn)) {
      continue;
    }

    // projen.web.ReactProject -> web.ReactProject
    const typename = fqn.substring(fqn.indexOf('.') + 1);

    const docsurl = `https://github.com/projen/projen/blob/master/API.md#projen-${typename.toLocaleLowerCase()}`;
    let pjid = typeinfo.docs?.custom?.pjid ?? decamelize(typename).replace(/_project$/, '');
    result.push({
      moduleName: typeinfo.assembly,
      typename,
      pjid,
      fqn,
      options: discoverOptions(jsii, fqn).sort((o1, o2) => o1.name.localeCompare(o2.name)),
      docs: typeinfo.docs?.summary,
      docsurl,
    });
  }

  return result.sort((r1, r2) => r1.pjid.localeCompare(r2.pjid));
}

function discoverOptions(jsii: JsiiTypes, fqn: string): ProjectOption[] {
  const options: { [name: string]: ProjectOption } = {};
  const params = jsii[fqn]?.initializer?.parameters ?? [];
  const optionsParam = params[0];
  const optionsTypeFqn = optionsParam?.type?.fqn;

  if (params.length > 1 || (params.length === 1 && optionsParam?.name !== 'options')) {
    throw new Error(`constructor for project ${fqn} must have a single "options" argument of a struct type. got ${JSON.stringify(params)}`);
  }

  addOptions(optionsTypeFqn);

  const opts = Object.values(options);

  return opts.sort((a, b) => a.switch.localeCompare(b.switch));

  function addOptions(ofqn?: string, basePath: string[] = [], optional = false) {
    if (!ofqn) {
      return;
    }

    const struct = jsii[ofqn];
    if (!struct) {
      throw new Error(`unable to find options type ${ofqn} for project ${fqn}`);
    }

    for (const prop of struct.properties ?? []) {
      const propPath = [...basePath, prop.name];

      // protect against double-booking
      if (prop.name in options) {
        throw new Error(`duplicate option "${prop.name}" in ${fqn} (already declared in ${options[prop.name].parent})`);
      }

      let typeName;
      if (prop.type?.primitive) {
        typeName = prop.type?.primitive; // e.g. 'string', 'boolean', 'number'
      } else if (prop.type?.fqn) {
        typeName = prop.type?.fqn.split('.').pop(); // projen.NodeProjectOptions -> NodeProjectOptions
      } else { // any other types such as collection types
        typeName = 'unknown';
      }

      const isOptional = optional || prop.optional;
      let defaultValue = prop.docs?.default;

      if (defaultValue === 'undefined') {
        defaultValue = undefined;
      }

      // if this is a mandatory option and we have a default value, it has to be JSON-parsable to the correct type
      if (!isOptional && defaultValue) {
        if (!prop.type?.primitive) {
          throw new Error(`required option "${prop.name}" with a @default must use primitive types (string, number or boolean). type found is: ${typeName}`);
        }

        checkDefaultIsParsable(prop.name, defaultValue, prop.type?.primitive);
      }

      options[prop.name] = filterUndefined({
        path: propPath,
        parent: struct.name,
        name: prop.name,
        docs: prop.docs.summary,
        type: typeName,
        switch: propPath.map(p => decamelize(p).replace(/_/g, '-')).join('-'),
        default: defaultValue,
        optional: isOptional,
        deprecated: prop.docs.stability === 'deprecated' ? true : undefined,
      });
    }

    for (const ifc of struct.interfaces ?? []) {
      addOptions(ifc);
    }
  }
}

function filterUndefined(obj: any) {
  const ret: any = {};
  for (const [k, v] of Object.entries(obj)) {
    if (v !== undefined) {
      ret[k] = v;
    }
  }
  return ret;
}

function isProjectType(jsii: JsiiTypes, fqn: string) {
  const type = jsii[fqn];

  if (type.kind !== 'class') {
    return false;
  }
  if (type.abstract) {
    return false;
  }

  if (type.docs?.deprecated) {
    return false;
  }

  let curr = type;
  while (true) {
    if (curr.fqn === PROJECT_BASE_FQN) {
      return true;
    }

    if (!curr.base) {
      return false;
    }

    curr = jsii[curr.base];
    if (!curr) {
      return false;
    }
  }
}

function checkDefaultIsParsable(prop: string, value: string, type: string) {
  // macros are pass-through
  if (value.startsWith('$')) {
    return;
  }
  try {
    const parsed = JSON.parse(value);
    if (typeof(parsed) !== type) {
      throw new Error(`cannot parse @default value for mandatory option ${prop} as a ${type}: ${parsed}`);
    }

  } catch (e) {
    throw new Error(`unable to JSON.parse() value "${value}" specified as @default for mandatory option "${prop}": ${e.message}`);
  }
}