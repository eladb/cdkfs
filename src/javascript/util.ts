import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
