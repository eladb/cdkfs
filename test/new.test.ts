// tests for `projen new`: we run `projen new` for each supported project type
// and compare against a golden snapshot.
import { execSync, spawnSync } from 'child_process';
import { join } from 'path';
import { mkdirSync, readFileSync, removeSync } from 'fs-extra';
import { PROJEN_RC } from '../src/common';
import * as inventory from '../src/inventory';
import { mkdtemp, synthSnapshot, synthSnapshotWithPost, TestProject } from './util';

const cli = require.resolve('../bin/projen');

for (const type of inventory.discover()) {
  test(`projen new ${type.pjid}`, () => {
    const outdir = mkdtemp();
    try {
      const projectdir = createProjectDir(outdir);

      // execute `projen new PJID --no-synth` in the project directory
      const ret = spawnSync(process.execPath, [cli, 'new', '--no-synth', type.pjid], { cwd: projectdir });

      // throw on errors
      if (ret.error) { throw new Error(ret.error.message); }
      if (ret.status !== 0) { throw new Error(ret.stderr.toString('utf-8')); }

      // compare generated .projenrc.js to the snapshot
      const projenrc = readFileSync(join(projectdir, PROJEN_RC), 'utf-8');
      expect(projenrc).toMatchSnapshot();
    } finally {
      removeSync(outdir);
    }
  });
}

test('post-synthesis option enabled', () => {
  const project = new TestProject();

  expect(synthSnapshotWithPost(project)['.postsynth']).toContain('postsynth');
});

test('post-synthesis option disabled', () => {
  const project = new TestProject();

  expect(synthSnapshot(project)['.postsynth']).toBeUndefined();
});

function createProjectDir(workdir: string) {
  // create project under "my-project" so that basedir is deterministic
  const projectdir = join(workdir, 'my-project');
  mkdirSync(projectdir);

  const git = (command: string) => execSync(`git ${command}`, { cwd: projectdir });
  git('init');
  git('remote add origin git@boom.com:foo/bar.git');
  git('config user.name "My User Name"');
  git('config user.email "my@user.email.com"');
  return projectdir;
}