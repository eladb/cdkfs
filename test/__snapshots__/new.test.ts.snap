// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creating java project with enum-typed CLI arg 1`] = `
"import io.github.cdklabs.projen.java.JavaProject;
import io.github.cdklabs.projen.java.JavaProjectOptions;
import io.github.cdklabs.projen.ProjectType;

public class projenrc {
    public static void main(String[] args) {
        JavaProject project = new JavaProject(JavaProjectOptions.builder()
            .artifactId(\\"my-app\\")
            .groupId(\\"org.acme\\")
            .name(\\"my-project\\")
            .version(\\"0.1.0\\")
            .projectType(ProjectType.LIB)
            .projenrcJava(true)
            .build());
        project.synth();
    }
}"
`;

exports[`creating node project with enum-typed CLI arg 1`] = `
"const { javascript } = require(\\"projen\\");
const project = new javascript.NodeProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  packageManager: javascript.NodePackageManager.NPM,

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();"
`;

exports[`creating python project with enum-typed CLI arg 1`] = `
"from projen.python import PythonProject
from projen import ProjectType

project = PythonProject(
    author_email=\\"my@user.email.com\\",
    author_name=\\"My User Name\\",
    module_name=\\"my_project\\",
    name=\\"my-project\\",
    version=\\"0.1.0\\",
    project_type=ProjectType.LIB,
    projenrc_python=True,
);

project.synth();"
`;

exports[`projen new awscdk-app-java 1`] = `
Object {
  "src/test/java/projenrc.java": "import io.github.cdklabs.projen.awscdk.AwsCdkJavaApp;
import io.github.cdklabs.projen.awscdk.AwsCdkJavaAppOptions;

public class projenrc {
    public static void main(String[] args) {
        AwsCdkJavaApp project = new AwsCdkJavaApp(AwsCdkJavaAppOptions.builder()
            .artifactId(\\"my-app\\")
            .cdkVersion(\\"^1.130.0\\")
            .groupId(\\"org.acme\\")
            .mainClass(\\"org.acme.App\\")
            .name(\\"my-project\\")
            .version(\\"0.1.0\\")
            .build());
        project.synth();
    }
}",
}
`;

exports[`projen new awscdk-app-ts 1`] = `
Object {
  ".projenrc.js": "const { awscdk } = require(\\"projen\\");
const project = new awscdk.AwsCdkTypeScriptApp({
  cdkVersion: \\"1.129.0\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // cdkDependencies: undefined,  /* Which AWS CDK modules (those that start with \\"@aws-cdk/\\") this app uses. */
  // deps: [],                    /* Runtime dependencies of this module. */
  // description: undefined,      /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],                 /* Build dependencies for this module. */
  // packageName: undefined,      /* The \\"name\\" in package.json. */
  // release: undefined,          /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new awscdk-construct 1`] = `
Object {
  ".projenrc.js": "const { awscdk } = require(\\"projen\\");
const project = new awscdk.AwsCdkConstructLibrary({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  cdkVersion: \\"2.0.0\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new cdk8s-app-ts 1`] = `
Object {
  ".projenrc.js": "const { cdk8s } = require(\\"projen\\");
const project = new cdk8s.Cdk8sTypeScriptApp({
  cdk8sVersion: \\"1.0.0-beta.10\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new cdk8s-construct 1`] = `
Object {
  ".projenrc.js": "const { cdk8s } = require(\\"projen\\");
const project = new cdk8s.ConstructLibraryCdk8s({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  cdk8sVersion: \\"1.0.0-beta.10\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new cdktf-construct 1`] = `
Object {
  ".projenrc.js": "const { cdktf } = require(\\"projen\\");
const project = new cdktf.ConstructLibraryCdktf({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  cdktfVersion: \\"0.4.0\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new java 1`] = `
Object {
  "src/test/java/projenrc.java": "import io.github.cdklabs.projen.java.JavaProject;
import io.github.cdklabs.projen.java.JavaProjectOptions;

public class projenrc {
    public static void main(String[] args) {
        JavaProject project = new JavaProject(JavaProjectOptions.builder()
            .artifactId(\\"my-app\\")
            .groupId(\\"org.acme\\")
            .name(\\"my-project\\")
            .version(\\"0.1.0\\")
            .build());
        project.synth();
    }
}",
}
`;

exports[`projen new jsii 1`] = `
Object {
  ".projenrc.js": "const { cdk } = require(\\"projen\\");
const project = new cdk.JsiiProject({
  author: \\"My User Name\\",
  authorAddress: \\"my@user.email.com\\",
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  repositoryUrl: \\"git@boom.com:foo/bar.git\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new nextjs 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.NextJsProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
  // tailwind: true,          /* Setup Tailwind CSS as a PostCSS plugin. */
});
project.synth();",
}
`;

exports[`projen new nextjs-ts 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.NextJsTypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
  // tailwind: true,          /* Setup Tailwind CSS as a PostCSS plugin. */
});
project.synth();",
}
`;

exports[`projen new node --outdir path/to/mydir 1`] = `
"const { javascript } = require(\\"projen\\");
const project = new javascript.NodeProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();"
`;

exports[`projen new node 1`] = `
Object {
  ".projenrc.js": "const { javascript } = require(\\"projen\\");
const project = new javascript.NodeProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new project 1`] = `Object {}`;

exports[`projen new python 1`] = `
Object {
  ".projenrc.js": "const { python } = require(\\"projen\\");
const project = new python.PythonProject({
  authorEmail: \\"my@user.email.com\\",
  authorName: \\"My User Name\\",
  moduleName: \\"my_project\\",
  name: \\"my-project\\",
  version: \\"0.1.0\\",

  // deps: [],                /* List of runtime dependencies for this project. */
  // description: undefined,  /* A short description of the package. */
  // devDeps: [],             /* List of dev dependencies for this project. */
  // pip: true,               /* Use pip with a requirements.txt file to track project dependencies. */
  // poetry: false,           /* Use poetry to manage your project dependencies, virtual environment, and (optional) packaging/publishing. */
  // pytest: true,            /* Include pytest tests. */
  // setuptools: undefined,   /* Use setuptools with a setup.py script for packaging and publishing. */
  // venv: true,              /* Use venv to manage a virtual environment for installing dependencies inside. */
});
project.synth();",
}
`;

exports[`projen new react 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.ReactProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new react-ts 1`] = `
Object {
  ".projenrc.js": "const { web } = require(\\"projen\\");
const project = new web.ReactTypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new typescript 1`] = `
Object {
  ".projenrc.js": "const { typescript } = require(\\"projen\\");
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projen new typescript-app 1`] = `
Object {
  ".projenrc.js": "const { typescript } = require(\\"projen\\");
const project = new typescript.TypeScriptAppProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();",
}
`;

exports[`projenrc-json creates node-project 1`] = `
Object {
  "defaultReleaseBranch": "main",
  "name": "my-project",
  "projenrcJson": true,
  "type": "projen.javascript.NodeProject",
}
`;

exports[`projenrc-ts creates typescript projenrc 1`] = `
"import { typescript } from \\"projen\\";
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: \\"main\\",
  name: \\"my-project\\",
  projenrcTs: true,

  // deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],             /* Build dependencies for this module. */
  // packageName: undefined,  /* The \\"name\\" in package.json. */
  // release: undefined,      /* Add release management to this project. */
});
project.synth();"
`;
